cmd/serve/main.go

```diff
@@ -1,22 +1,45 @@
 package main
 
 import (
+	"context"
 	"embed"
 	"fmt"
 	"io/fs"
 	"log"
 	"net/http"
+	"os"
+
+	"github.com/jmoiron/sqlx"
+	_ "github.com/lib/pq"
+
+	handler_v1 "github.com/sunb26/xat/handler"
 )
 
 //go:embed all:web all:web/_next
 var content embed.FS
 
+func injectDB(db *sqlx.DB, next http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := context.WithValue(r.Context(), "db", db)
+
+		next.ServeHTTP(w, r.WithContext(ctx))
+	}
+}
+
 func main() {
 	content, err := fs.Sub(content, "web")
 	if err != nil {
 		log.Fatal(err)
 	}
+
+	db, err := sqlx.Open("postgres", os.Getenv("DSN"))
+	if err != nil {
+		log.Fatalf("failed to open to database: %v", err)
+	}
+
+	http.HandleFunc("/api/v1/user", injectDB(db, handler_v1.CreateUser))
 	http.Handle("/", http.FileServerFS(content))
+
 	fmt.Println("Listening on 127.0.0.1:3000")
 	log.Fatal(http.ListenAndServe(":3000", nil))
```

Comment:
right now we're using the default mux, which is why `nil` is passed on this line

regarding how middleware is structured, perhaps it would be cleaner to apply it to the mux directly like https://arc.net/l/quote/fapiwpmz. since you can attach multiple handlers to a mux, the mux is like an endpoint grouping tool allowing you to apply transforms on multiple related endpoints. iirc, mux can also register other mux, which allows for grouping of routes and collecting those muxes to pass to `ListenAndServe`

Reply:
Wow actually pretty cool stuff

Reply:
maybe put the static file server into its own mux? b/c it doesn't need a database connection in the context to serve static files. also suggestion to change `injector` to `middleware` instead, since adding a variable to context does not really consistute dependency injection